### PR TITLE
Nickname ui

### PR DESCRIPTION
**Pull request by shelbaz**
Adding nickname UI, logic, unit and espresso test (not done yet)

### CODE REVIEW

File: src/org/thoughtcrime/securesms/NicknameHandler.java [hunk 1]
```diff
@@ -25,7 +25,7 @@ public NicknameHandler setContext(Context context) {
      * This method is implemented for only testing purposes inorder
      * to pass mocked database object
      * @param recipientDatabase
-     * @return
+     * @rezturn
```
Comment (erinb90) [inline src/org/thoughtcrime/securesms/NicknameHandler.java]: rezturn?

File: src/org/thoughtcrime/securesms/NicknameHandler.java [hunk 2]
```diff
@@ -43,14 +43,15 @@ public NicknameHandler setRecipientDatabase(RecipientDatabase recipientDatabase)
      */
     public boolean setNickname(Recipient recipient, String nickname) {
         this.setupDatabaseHandler();
-        return this.recipientDatabase.setNickname(recipient, nickname);
+        try{this.recipientDatabase.setNickname(recipient, nickname);}
```
Comment (erinb90) [inline src/org/thoughtcrime/securesms/NicknameHandler.java]: formatting (also for the method below this)

File: src/org/thoughtcrime/securesms/NicknameHandler.java (hunk 1 above)
Comment (g-harel) [reply src/org/thoughtcrime/securesms/NicknameHandler.java]: new feature in Java X, such courage

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java [hunk 3]
```diff
@@ -61,6 +61,18 @@ public ConversationsHelper goConversations() {
         return new ConversationsHelper(new HelperSecret());
     }
 
+    public PreferencesHelper goSettings() {
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: nooooooooooooooooooooooooooooooo

Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: andddd its gone. forgot to remove it

File: res/xml/recipient_preferences.xml [hunk 4]
```diff
@@ -33,6 +33,16 @@
                 android:positiveButtonText="@null"
                 android:persistent="false"
                 app:numColumns="5" />
+        <EditTextPreference
+            android:defaultValue="Username"
```
Comment (g-harel) [inline res/xml/recipient_preferences.xml]: This should be a string resource

File: res/xml/recipient_preferences.xml [hunk 5]
```diff
@@ -33,6 +33,16 @@
...
+            android:key="pref_key_change_nickname"
+            android:selectAllOnFocus="true"
+            android:singleLine="true"
+            android:title="Set Nickname" />
```
Comment (g-harel) [inline res/xml/recipient_preferences.xml]: This should be a string resource

File: res/xml/recipient_preferences.xml [hunk 6]
```diff
@@ -33,6 +33,16 @@
...
+        <Preference
+            android:key="pref_key_reset_nickname"
+            android:persistent="false"
+            android:title="Reset Nickname" />
```
Comment (g-harel) [inline res/xml/recipient_preferences.xml]: This should be a string resource

File: src/org/thoughtcrime/securesms/NicknameHandler.java [hunk 7]
```diff
@@ -24,7 +24,7 @@ public NicknameHandler setContext(Context context) {
      * to pass mocked database object
      *
      * @param recipientDatabase
-     * @return
+     * @rezturn
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/NicknameHandler.java]: ...

File: src/org/thoughtcrime/securesms/NicknameHandler.java [hunk 8]
```diff
@@ -42,7 +42,14 @@ public NicknameHandler setRecipientDatabase(RecipientDatabase recipientDatabase)
      */
     public boolean setNickname(Recipient recipient, String nickname) {
         this.setupDatabaseHandler();
-        return this.recipientDatabase.setNickname(recipient, nickname);
+        try{ return this.recipientDatabase.setNickname(recipient, nickname);}
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/NicknameHandler.java]: this block is very hard to read.

File: src/org/thoughtcrime/securesms/NicknameHandler.java [hunk 9]
```diff
@@ -42,7 +42,14 @@ public NicknameHandler setRecipientDatabase(RecipientDatabase recipientDatabase)
...
+        catch (Exception e) { System.err.println("Did not add nickname"); return false;}
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/NicknameHandler.java]: It's probably not a great Idea to silently ignore exceptions

File: src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java [hunk 10]
```diff
@@ -193,6 +197,7 @@ private void setHeader(@NonNull Recipient recipient) {
 
     this.avatar.setBackgroundColor(recipient.getColor().toActionBarColor(this));
     this.toolbarLayout.setTitle(recipient.toShortString());
+
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java]: ?

File: src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java [hunk 11]
```diff
@@ -106,7 +111,6 @@ public void onCreate(Bundle instanceState, @NonNull MasterSecret masterSecret) {
     this.address       = getIntent().getParcelableExtra(ADDRESS_EXTRA);
 
     Recipient recipient = Recipient.from(this, address, true);
-
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java]: ?

File: src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java [hunk 12]
```diff
@@ -496,6 +508,48 @@ protected Void doInBackground(Void... params) {
       }
     }
 
+    private class NicknameChangeListener implements Preference.OnPreferenceChangeListener {
+
+      @Override
+      public boolean onPreferenceChange(Preference preference, Object newValue) {
+        NicknameHandler nicknameHandler = new NicknameHandler(getContext());
+
+        try{nicknameHandler.setNickname(recipient, newValue.toString());}
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java]: Same feedback as earlier. Also, what error are you expecting? Could the catch be made more specific?

File: src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java [hunk 13]
```diff
@@ -496,6 +508,48 @@ protected Void doInBackground(Void... params) {
...
+        Toast toast=Toast.makeText(getContext(), outputMessage, Toast.LENGTH_SHORT);
+        toast.show();
+      }
+    }
+
+    private class NicknameResetClickedListener implements Preference.OnPreferenceClickListener {
+      @Override
+      public boolean onPreferenceClick(Preference preference) {
+        NicknameHandler nicknameHandler = new NicknameHandler(getContext());
+        try{nicknameHandler.removeNickname(recipient);}
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/RecipientPreferenceActivity.java]: same

File: src/org/thoughtcrime/securesms/database/RecipientDatabase.java [hunk 14]
```diff
@@ -243,7 +243,12 @@ public void setBlocked(@NonNull Recipient recipient, boolean blocked) {
   public boolean setNickname(@Nullable Recipient recipient, String nickname) {
     ContentValues values = new ContentValues();
     values.put(NICKNAME, nickname);
-    return processNicknameSqlRequest(values, recipient.getAddress()) > 0;
+    int res = processNicknameSqlRequest(values, recipient.getAddress());
+    if (res > 0) {
+      recipient.resolve().setNickname(nickname);
+      return true;
+    }
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/database/RecipientDatabase.java:250]: you don't need the else there.

File: src/org/thoughtcrime/securesms/recipients/Recipient.java [hunk 15]
```diff
@@ -117,8 +118,10 @@ public static void applyCached(@NonNull Address address, Consumer<Recipient> con
     this.color                = null;
     this.resolving            = true;
 
+
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java]: ?

File: src/org/thoughtcrime/securesms/recipients/Recipient.java [hunk 16]
```diff
@@ -264,6 +267,8 @@ public void setContactUri(@Nullable Uri contactUri) {
       return Util.join(names, ", ");
     }
 
+    if (this.nickname != null) return this.nickname;
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java]: The pattern in this project has generally been to write the block `if () {...` so I think that should be followed

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java [hunk 17]
```diff
@@ -105,4 +105,11 @@ public PinnedHelper goPinned() {
 
         return new PinnedHelper(new HelperSecret());
     }
+
+    public PreferencesHelper goSettings() {
+        onView(withId(R.id.contact_photo_image))
+                .perform(click());
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: indentation is off

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java [hunk 18]
```diff
@@ -14,4 +14,49 @@ public Helper(ActivityTestRule<ConversationListActivity> activityRule) {
     public ConversationsHelper goConversations() {
         return new ConversationsHelper(new HelperSecret());
     }
+
+    public PreferencesHelper resetNickname() {
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: this method should be on the PreferencesHelper class

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java [hunk 19]
```diff
@@ -14,4 +14,49 @@ public Helper(ActivityTestRule<ConversationListActivity> activityRule) {
...
+        return new PreferencesHelper(new HelperSecret());
+    }
+
+    public PreferencesHelper setNickname(String message) {
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: should also be in PreferencesHelper class. Helper is extended by helpers for all pages ...

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java [hunk 20]
```diff
@@ -14,4 +14,49 @@ public Helper(ActivityTestRule<ConversationListActivity> activityRule) {
...
+        return new PreferencesHelper(new HelperSecret());
+    }
+
+    /* ASSERTIONS */
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: All of these assertions are already in BaseHelper, I imagine this is from the merge, but please remove em all.

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java [hunk 21]
```diff
@@ -14,4 +14,49 @@ public Helper(ActivityTestRule<ConversationListActivity> activityRule) {
...
+
+            throw new Error("Helper.assertNoText: View found matching \"" + text + "\"");
+        } catch (NoMatchingViewException e) {}
+
+        return (T)this;
+    }
+
+    /* MISC */
+
+    public String randString() {
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: same, this is in BaseHelper now

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+import static android.support.test.espresso.action.ViewActions.typeText;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+
+public class PreferencesHelper extends Helper<PreferencesHelper>{
+    public PreferencesHelper(HelperSecret s) {}
+
+    public PreferencesHelper resetNickname() {
+        onView(withText("Reset Nickname"))
+                .perform(click());
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java]: formatting is off

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+import static android.support.test.espresso.action.ViewActions.replaceText;
+import static android.support.test.espresso.action.ViewActions.typeText;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+
+public class PreferencesHelper extends Helper<PreferencesHelper>{
+    public PreferencesHelper(HelperSecret s) {}
+
+    public PreferencesHelper resetNickname() {
+        onView(withText("Reset Nickname"))
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java]: this should be a string resource.

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+                .perform(click());
+
+        return new PreferencesHelper(new HelperSecret());
+    }
+
+    public PreferencesHelper setNickname(String message) {
+        onView(withText("Set Nickname"))
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java]: this should be a string resource.

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+                .perform(click());
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/PreferencesHelper.java]: formatting is off on these three lines.

File: test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java [hunk 26]
```diff
@@ -14,6 +16,8 @@
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.mockingDetails;
 import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.spy;
+
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java]: extra line.

File: test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java [hunk 27]
```diff
@@ -31,10 +35,10 @@ public void setUp() throws Exception {
         recipientDatabase = mock(RecipientDatabase.class);
     }
 
+
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java]: ?

File: test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java [hunk 28]
```diff
@@ -31,10 +35,10 @@ public void setUp() throws Exception {
...
     protected void setupSetNicknameMethod() {
-        when(recipientDatabase.setNickname(recipient, "test")).thenReturn(true)
+        when(recipientDatabase.setNickname(recipient, "test" )).thenReturn(true)
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java]: ?

**Review by g-harel — CHANGES_REQUESTED** [review]

File: src/org/thoughtcrime/securesms/recipients/Recipient.java [hunk 29]
```diff
@@ -468,6 +477,14 @@ public void setRingtone(@Nullable Uri ringtone) {
     notifyListeners();
   }
 
+  public void setNickname(String nickname) {
+    synchronized (this) {
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java:480]: is there a reason synchronized is not on the method (like for the others)

File: test/androidTest/java/org/thoughtcrime/securesms/NicknameEspressoTest.java [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class NicknameEspressoTest {
+    @Rule
+    public ActivityTestRule<ConversationListActivity> mainActivityRule =
+            new ActivityTestRule(ConversationListActivity.class, true, false);
+
+    @Test
+    public void pageExists(){
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/NicknameEspressoTest.java]: `() {`

File: src/org/thoughtcrime/securesms/recipients/Recipient.java (hunk 29 above)
Comment (shelbaz) [reply src/org/thoughtcrime/securesms/recipients/Recipient.java:480]: That's the convention for the setters, check the rest of the file.

Comment (g-harel) [reply src/org/thoughtcrime/securesms/recipients/Recipient.java:480]: :+1:

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java (hunk 17 above)
Comment (shelbaz) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: Isn't this the convention u set?...

Comment (g-harel) [reply test/androidTest/java/org/thoughtcrime/securesms/espresso/ConversationHelper.java]: not 8 spaces

File: test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java [hunk 31]
```diff
@@ -14,4 +14,8 @@ public Helper(ActivityTestRule<ConversationListActivity> activityRule) {
     public ConversationsHelper goConversations() {
         return new ConversationsHelper(new HelperSecret());
     }
+
+    /* ASSERTIONS */
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/espresso/Helper.java]: this comment is now useless

File: src/org/thoughtcrime/securesms/NicknameHandler.java (hunk 1 above)
Comment (shelbaz) [reply src/org/thoughtcrime/securesms/NicknameHandler.java]: prolly click command z somewhere and didnt notice the change

**Comment by shelbaz** [comment]
Changes done

**Review by g-harel — APPROVED** [review]

**Review by Radu-Raicea — APPROVED** [review]

**Pull request merged**